From ac8ada79443153c754eaa65e513ccd3b19959b2b Mon Sep 17 00:00:00 2001
From: Zhantao Tang <zhantao.tang@windriver.com>
Date: Sun, 16 May 2021 14:35:07 +0800
Subject: [PATCH] sw: change kmalloc flag to GFP_ATOMIC to fix debug calltrace

Issue: LINCD-5401

When setting ip address for pfe ports by ifconfig/dhcp, the rx mode
will be set and netif_addr_lock_bh() will be called to into atomic
context, while when setting rx mode, pfeng_netif_add_mac_to_request_list()
will be called, and if the mac address is new for the request list,
kmalloc() will allocate new memory with GFP_KERNEL flag, so that it
may sleep, and when debug configs enabled, the following calltrace
will show:

	[ 191.943726] Call trace:
	[ 191.946161] dump_backtrace+0x0/0x1d4
	[ 191.949803] show_stack+0x24/0x30
	[ 191.953105] dump_stack+0xf0/0x13c
	[ 191.956490] ___might_sleep+0x1a4/0x244
	[ 191.960305] __might_sleep+0x60/0x9c
	[ 191.963870] kmem_cache_alloc_trace+0x318/0x47c
	[ 191.968449] pfeng_netif_add_mac_to_request_list+0x9c/0x120 [pfeng]
	[ 191.974671] pfeng_netif_logif_set_rx_mode+0xac/0x230 [pfeng]
	[ 191.980361] __dev_set_rx_mode+0x60/0xa0
	[ 191.984264] dev_mc_add+0x9c/0xc0
	[ 191.987563] igmp6_group_added+0x118/0x144
	[ 191.991641] __ipv6_dev_mc_inc+0x21c/0x3fc
	[ 191.995720] __ipv6_sock_mc_join+0x1a8/0x3c4
	[ 191.999973] ipv6_sock_mc_join+0x20/0x2c
	[ 192.003882] do_ipv6_setsockopt.constprop.0+0x1150/0x1330
	[ 192.009261] ipv6_setsockopt+0x80/0x1d0
	[ 192.013082] udpv6_setsockopt+0x28/0x50
	[ 192.016903] sock_common_setsockopt+0x28/0x34
	[ 192.021241] __sys_setsockopt+0xa4/0x16c
	[ 192.025147] __arm64_sys_setsockopt+0x34/0x44
	[ 192.029489] el0_svc_common.constprop.0+0x9c/0x174
	[ 192.034260] do_el0_svc+0x78/0x9c
	[ 192.037558] el0_svc+0x20/0x30
	[ 192.040596] el0_sync_handler+0x1a4/0x1b0
	[ 192.044589] el0_sync+0x178/0x180

So chage the kmalloc flag to GFP_ATOMIC to fix the issue.

Upstream-Status: Pengding

Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 sw/linux-pfeng/pfeng-netif.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/sw/linux-pfeng/pfeng-netif.c b/sw/linux-pfeng/pfeng-netif.c
index 079f8b4..e4e5206 100644
--- a/sw/linux-pfeng/pfeng-netif.c
+++ b/sw/linux-pfeng/pfeng-netif.c
@@ -687,7 +687,7 @@ static int pfeng_netif_add_mac_to_request_list(struct net_device *netdev, struct
 
 	if(!is_present) {
 		/* Add address to the list */
-		entry = kmalloc(sizeof(pfeng_netif_mac_db_list_entry_t), GFP_KERNEL);
+		entry = kmalloc(sizeof(pfeng_netif_mac_db_list_entry_t), GFP_ATOMIC);
 		if (!entry) {
 			netdev_err(netdev, "Memory allocation failed\n");
 			ret = ENOMEM;
-- 
2.25.1

